### PR TITLE
RSDK-6992 RSDK-6993 - implement DataCollector pt. 1

### PR DESCRIPTION
**Pull request by gvaradarajan**
Part 1 of 2 PRs

The TODO comments are there to provide a sense of the direction of the architecture moving forward and will be removed before merge

For now, to run the tests introduced in the PR, add "data" to the default features in micro-rdk/Cargo.toml before calling `make test`

### CODE REVIEW

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:127]: Looks like the last sentence is incomplete. I was going to say a sentence fragement, but that isn't exactly true.

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:35]: Looks like these cases are already handled via `ok_or` below?

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:92]: If the only real use of this is for `fmt`, maybe just inline it in there rather than synthesizing a new method name?

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:78]: This design is very centralized / closed / static, in that if a new value returning method was added to one of these component types, we would need to make a compensating change here before it could be used with data capture. I'm not necessarily saying that's a bad design: it makes it a lot easier to see what is going on. The other way to do this would be to build into each component some ability to enumerate the method names / return Fn objects for each. That'd result in a more decentralized and open design. That said, nothing prevents us from moving to such a design later. So I think on balance I'm fine with starting with the closed design.

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:51]: I think maybe remove the commented out code under the `TODO` sections (here and elsewhere) in favor of just a description. The code is unlikely to remain useful and correct because as things change under review these probably won't keep up.

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:159]: Or just `1000.0 / capture_frequence_hz as u64`

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:193]: `self.method.to_string()`

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [reply micro-rdk/src/common/data_collector.rs:92]: And anyway don't you get `ToString` for free if you implement `Display`?

**Review by acmorrow — CHANGES_REQUESTED** [review]

File: micro-rdk/src/common/sensor.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/sensor.rs:94]: this was a mistake, I meant to initialize the `time_requested` before the call to `get_sensor_readings`

File: micro-rdk/src/common/data_collector.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/data_collector.rs:51]: sorry I think this was a miscommunication on my part, but it's my plan to delete the TODO comments before merge (although on reflection, I like keeping the comment and removing the commented-out code, thoughts?)

File: micro-rdk/src/common/data_collector.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/data_collector.rs:78]: Yeah I had considered that open design and I do agree that I like it better in theory, but I did not want to grapple with the lifetime considerations for function objects, so the design decision was born partially out of laziness

File: micro-rdk/src/common/data_collector.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/data_collector.rs:92]: both fair points

File: micro-rdk/src/common/data_collector.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/data_collector.rs:127]: yeah, should have backspaced harder

File: micro-rdk/src/common/data_collector.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/data_collector.rs:159]: yeah my bad, wrote the way I think and then forgot to clean up

File: micro-rdk/src/common/robot.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/robot.rs:67]: I was concerned I would get a dead code complaint, but we actually ignore those at the top of the file. Let me try removing it

File: micro-rdk/src/common/sensor.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/sensor.rs:69]: Yeah, fair this function isn't quite the answer, I'll implement your second suggestion

File: micro-rdk/src/common/sensor.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/sensor.rs:73]: After some experimentation and examination of the analogous code in RDK, I discovered that this is the format that the data management service on app understands how to parse for the readings collection method.

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [reply micro-rdk/src/common/data_collector.rs:46]: Keep the TODOs, with as much explication as you see fit, but drop the code would be my recommendation.

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [reply micro-rdk/src/common/data_collector.rs:73]: Laziness - one of the principle virtues of the programmer! I'm fine with the closed approach, for now.

File: micro-rdk/src/common/sensor.rs
Comment (acmorrow) [reply micro-rdk/src/common/sensor.rs]: Does it basically expect the method name to be played back as the key or something?

File: micro-rdk/src/common/sensor.rs
Comment (acmorrow) [reply micro-rdk/src/common/sensor.rs]: I think the same mistake is exists in the other place you do this too.

File: micro-rdk/src/common/sensor.rs
Comment (acmorrow) [inline micro-rdk/src/common/sensor.rs:82]: Wondering a bit about whether Metadata production belongs in the leaf functions here. If `get_readings_data` returned a `sensor_data::Data`, and `Vector3::to_sensor_data` the same, could timestamp management all be lifted up to `DataCollector`? It'd let you get, for instance, the timestamp fields out of the arguments to `Vector3::to_sensor_data`, which in turn would probably let you use `From` there as well.

The downside I see to that choice would be if the metadata later was expanded in a way that could only be satisfied by the objects themselves.

What do you think?

File: micro-rdk/src/common/sensor.rs
Comment (acmorrow) [inline micro-rdk/src/common/sensor.rs:26]: I think these can be combined.

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:1]: Is there actual dead code in this file? If so can this be scoped more narrowly (genuinely asking, as I don't know).

File: micro-rdk/src/common/data_collector.rs
Comment (acmorrow) [inline micro-rdk/src/common/data_collector.rs:169]: This `Local::now` is duplicated work in the `Sensor::get_readings_data` case

**Review by acmorrow — COMMENTED** [review]
Looking pretty good, with a design question.

File: micro-rdk/src/common/sensor.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/sensor.rs:82]: Originally my plan was to put this function in `DataCollector` rather than in this trait/ Then I could create `From<_> for SensorData` impls for both Vector3 and GenericSensorReadings that initially set `metadata` to None and leave it to the `DataCollector` to determine `metadata`. But the downside you mentioned came to mind and I decided against it. 

However, if you think the other design is better enough to be worth that risk, I'm happy to do that instead

File: micro-rdk/src/common/data_collector.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/data_collector.rs:1]: This was a mistake. It was only a problem if one moved "data" into the default features and ran `make clippy-esp32`. I'll remove

File: micro-rdk/src/common/sensor.rs
Comment (acmorrow) [reply micro-rdk/src/common/sensor.rs:83]: I'm leaning towards making the change. After all, the goal here is for the sensor to provide the information that goes into the `.data` `oneof`. It seems fair to conclude that the sensor shouldn't need to be involved in deciding any metadata that the data sync service might need in addition to the sensors own readings.

Also, I wonder if since `sensor_data::Data` is a real type, it wouldn't make more sense to create `From<_> for sensor_data::Data`? Then the metadata isn't even part of the story and there is no need to add any special handling around `None` for the `metadata` field.

**Review by acmorrow — APPROVED** [review]
This LGTM if you want to keep the current design, or, we can do another round if you decide to follow through on the alternative design.

File: micro-rdk/src/common/sensor.rs
Comment (gvaradarajan) [reply micro-rdk/src/common/sensor.rs:83]: yeah, it's better to write conversions to `sensor_data::Data`. Ok, I'll make the change

**Review by acmorrow — APPROVED** [review]
LGTM

File: micro-rdk/Cargo.toml
Comment (gvaradarajan) [reply micro-rdk/Cargo.toml]: yep lol

**Pull request merged**
